### PR TITLE
engine_rocks,engine_traits: abstract rocks util

### PR DESCRIPTION
**Pull request by Rustin-Liu**
###  What have you changed?
Abstract rocks util.
###  What is the type of the changes?

- Misc (other changes)

###  How is the PR tested?

- Unit test

###  Refer to a related PR or issue link (optional)
#6402

### CODE REVIEW

**Comment by brson** [comment]
Beyond that though, there is an additional issue that needs to be fixed at some point, but I think is fair to not do in this PR:

The `MetricsFlusher` is parameterized over two databases, and ultimately those two databases need to be allowed to be two _different types_ of databases.

That tells me that this is implemented at the wrong level of abstraction: instead of MetricsFlusher being a trait that is implemented by each database, MetricsFlusher really wants to be a concrete type that is implemented in engine_traits itself.

To do that will probably require a good bit of refactoring, and maybe adding some new abstractions to the engines.

I think let's not try to do that now.

File: components/engine_traits/src/engine_util/metrics_flusher.rs [hunk 1]
```diff
@@ -0,0 +1,11 @@
+// Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.
```
Comment (hicqu) [inline components/engine_traits/src/engine_util/metrics_flusher.rs]: Seems the submodule  `engine_util` is also unnecessary.

Comment (Rustin-Liu) [reply components/engine_traits/src/engine_util/metrics_flusher.rs]: removed.

**Comment by hicqu** [comment]
@Rustin-Liu thanks for your contribution! Please consider these things:
1. remove submodule `engine_util` in `engine_traits`;
2. `MetricsFlusher` in `engine_traits` should be a generic struct instead of a trait;
3. Add a trait method `flush_metrics(&self, name: &str)` in `engine_traits::KvEngine`.

For task 3 it needs more to do to avoid passing `shared_block_cache` to `flush_metrics`. I will send a issue to discuss it soon.

**Comment by Rustin-Liu** [comment]
@hicqu Already removed engine_util submodule.
As @brson  say, I think we can do it later.  Because I have just started and it may be too complicated for me.
So I think we can refactor it later.

**Comment by Rustin-Liu** [comment]
@sre-bot  /run-all-tests

**Comment by AndreMouche** [comment]
Please fix the conflicts, thank you

**Comment by Rustin-Liu** [comment]
PTAL~

File: cmd/src/server.rs [hunk 2]
```diff
@@ -10,9 +10,11 @@
 
 use crate::{setup::*, signal_handler};
 use engine::{
-    rocks::util::{metrics_flusher::MetricsFlusher, security::encrypted_env_from_cipher_file},
-    rocks::{self, util::metrics_flusher::DEFAULT_FLUSHER_INTERVAL},
+    rocks::util::security::encrypted_env_from_cipher_file,
+    rocks::{self},
```
Comment (hicqu) [inline cmd/src/server.rs]: s/{self}/self

**Comment by hicqu** [comment]
@Rustin-Liu I posted my last comment, please resolve it if you are free.

File: cmd/src/server.rs (hunk 2 above)
Comment (Rustin-Liu) [reply cmd/src/server.rs]: Fixed.

**Comment by hicqu** [comment]
@Rustin-Liu do you have any plan to do this: https://github.com/tikv/tikv/pull/6661#issuecomment-589554153 after the PR is merged?

**Comment by Rustin-Liu** [comment]
> @Rustin-Liu do you have any plan to do this: [#6661 (comment)](https://github.com/tikv/tikv/pull/6661#issuecomment-589554153) after the PR is merged?

I want do it after I finished the util module abstract.

**Comment by Rustin-Liu** [comment]
@sre-bot  /merge

**Comment by Rustin-Liu** [comment]
@you06 ping~

**Comment by you06** [comment]
@Rustin-Liu you cannot use this command to merge PRs from yourself. @hicqu PTAL.

**Comment by Rustin-Liu** [comment]
> @Rustin-Liu you cannot use this command to merge PRs from yourself. @hicqu PTAL.

But this pr already have can merge label.  Do you mean in TiKV this rule can not work?

**Comment by Rustin-Liu** [comment]
@sre-bot  /run-all-tests

**Comment by you06** [comment]
@Rustin-Liu The guy who merge the PR should be responsible for it, that's why we recommend one the reviewers to trigger merge event.

**Comment by Rustin-Liu** [comment]
> /run-all-tests

It seems CI broken, I try the test in another PR it also failed.

**Pull request merged**